### PR TITLE
Add Hookspec for providing sample data

### PR DESCRIPTION
**Pull request by tlambert03**
# Description
I'm unable to count the number of times that I've opened napari and then went looking for some sample data... So here's a proposal for a hookspec that lets plugins provide sample data.  As a builtin plugin, I've added support for scikit-image (it checks whether scikit-image is available, without importing, and adds the menu only it if it's there).  Here's the basic idea:

https://user-images.githubusercontent.com/1609449/112767899-325b6580-8fe7-11eb-9512-15eb3640dbc5.mov

The general hookspec is:
```python
@napari_hook_specification
def napari_provide_sample_data() -> Dict[
    str, Union[Union[Path, str], Callable[..., Iterable[LayerData]]]
]:
    ...
```

That is: plugins are asked to return a dict, where the keys are the name of the sample (as will be shown in the menu), and the values are either callables that return an iterable of LayerData tuples, or a string (that will be passed to `viewer.open`).

programmatically, you can access this api with:
```python
# e.g.
viewer.open_sample('scikit-image', 'coins')

from napari.plugins import available_samples  # returns all (plugin, sample) pairs
```

we can definitely change/discuss the API, but I wanted to get some general feedback before going much farther.

(as for scikit-image, this _doesn't_ check for pooch... but you get a reasonable error if it's not there... and we can also discuss how @jni feels about a plugin called "scikit-image" 😄)

## Type of change

- [x] New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: napari/plugins/_skimage_data.py
Comment (jni) [inline napari/plugins/_skimage_data.py:1]: Oh man, this file points to terrible patterns in the skimage data model! :sweat_smile: :confounded: (specifically: you can't just get the dir of the data module, and we don't provide any metadata programmatically.) Nice work fixing our stuff. :facepalm:

File: napari/plugins/hook_specifications.py
Comment (jni) [inline napari/plugins/hook_specifications.py]:
```suggestion
```

**Review by jni — APPROVED** [review]
I love this @tlambert03!

File: napari/plugins/_skimage_data.py
Comment (tlambert03) [reply napari/plugins/_skimage_data.py:1]: well... to be fair to scikit-image, it would be very hard to avoid manual curation here.  For instance, `download_all` is a very reasonable thing to export... but not reasonable here.  and it's probably worth having `cat` and `chelsea` for historical reasons, but maybe not here...

**Comment by jni** [comment]
... Now looking at the list I notice we don't have the famous Fiji mitosis sample dataset. Would be a good one to have.

**Comment by tlambert03** [comment]
> I love this @tlambert03!

great!  are you even ok with the new method on `Viewer`: `viewer.load_sample_data`?  That was the main thing I thought might require discussion.

> Now looking at the list I notice we don't have the famous Fiji mitosis sample dataset. 

Yeah... it could be fun to fill out the builtin samples :)  we can definitely add more (and the ability to add from a string/URL is nice)

**Comment by jni** [comment]
> great! are you even ok with the new method on Viewer: viewer.load_sample_data? That was the main thing I thought might require discussion.

I did pause there, but I couldn't come up with a better idea. :joy:

**Comment by sofroniewn** [comment]
Just seeing this now, looks amazing! I'd love to play with it and give it a review before merge. I should be able to get to it soonish. Very exciting

File: napari/components/viewer_model.py
Comment (sofroniewn) [inline napari/components/viewer_model.py]: I think i prefer `open_sample_data` as it is more consistent with `open_sample`, but I am curious about alternative APIs where we try and just use open.

We already have the `path` and `plugin` kwargs - could we just use those? maybe use some syntax like `sample://coins` or `sample://scikit-image/coins`?

I think it would be nice if we didn't have to add another method? Though if people prefer to new method that's fine too. Then maybe `def open_sample(self, name: str, plugin: str)`

File: napari/components/viewer_model.py
Comment (sofroniewn) [inline napari/components/viewer_model.py:717]: Do we need the `**kwargs` here? I know we have them in `open`, but that seems like a bit of a different use case

File: napari/plugins/__init__.py
Comment (sofroniewn) [inline napari/plugins/__init__.py]: this copy is shallow right, it doesn't copy the underlying data, just the dict?

File: napari/plugins/_skimage_data.py
Comment (sofroniewn) [inline napari/plugins/_skimage_data.py]: minor, but can you use `name` instead of `n`

File: napari/plugins/_skimage_data.py
Comment (sofroniewn) [inline napari/plugins/_skimage_data.py:42]: @jni does the skimage data now use pooch? is it possible that might get pulled out one day into it's own thing?

I guess we were saying that once skimage drops matplotlib we might want to add it back as a dependency?

I like the idea of making this always available to users

**Review by sofroniewn — COMMENTED** [review]
Overall love it @tlambert03, I've left some minor comments / suggestions around the viewer method and a couple other things. The PR also needs some doc strings (i saw your comment noting that it wasn't done yet, but i'm just noting that so i don't forget). Otherwise I think it'll be good to go soon!! Great idea

I think this will be great for some of our new segmentation plugin developers to play with too cc @neuromusic

File: napari/plugins/_skimage_data.py
Comment (sofroniewn) [reply napari/plugins/_skimage_data.py]: i was wondering what an integer would be used for :-)

File: napari/plugins/_skimage_data.py
Comment (jni) [reply napari/plugins/_skimage_data.py:42]: yes we use pooch. Indeed, if we merge this, it might be good to add `scikit-image[data]` to `napari[all]`. People who want a more minimal install can always do `napari[pyqt]`?

File: napari/components/viewer_model.py
Comment (jni) [reply napari/components/viewer_model.py]: I like `open_sample`. The URL-style syntax I'm ok with, but I don't love it. (And yes it should be nested in any case)

File: napari/components/viewer_model.py
Comment (tlambert03) [reply napari/components/viewer_model.py:717]: this would actually let the data-getting function accept parameters (#1211).  imagine, for instance, the shepp-logan phantom sample in imageJ, that lets you give dimensions... or perhaps a PSF generator.  The visual representation of that would still need some work/discussion, but it would let you do `viewer.open_sample('plugin', 'sample', nx=512, dz=0.13)` ... or something.

File: napari/components/viewer_model.py
Comment (tlambert03) [reply napari/components/viewer_model.py]: I also like  `def open_sample(self, plugin: str, name: str)`, don't love the url syntax either.

as one final thing to consider (if we don't want a new method)... the `plugin` parameter to `viewer.open` could also accept a tuple or something, then `path` would be optional iff used as `viewer.open(plugin=('scikit-image', 'cat'))` or something like that.

I think I kinda prefer `open_sample`

File: napari/plugins/__init__.py
Comment (tlambert03) [reply napari/plugins/__init__.py]: correct, but probably could even be removed

File: napari/plugins/_skimage_data.py
Comment (tlambert03) [reply napari/plugins/_skimage_data.py:42]: > might be good to add scikit-image[data] to napari[all]

:heart:

**Comment by tlambert03** [comment]
ok, updated the `[all]` extra to include `scikit-image[data]`.  changed it to `viewer.open_sample`.  Added docstrings to everything... and added a test.  (Also fixed a small bug in `viewer.open()` when passing a `Path` object).

The main thing requiring final approval is just whether we're ok with `viewer.open_sample(self, plugin: str, sample: str, **kwargs)`  (both in terms of the name/signature, and in terms of whether you'd prefer to leave `kwargs` for later discussion)

File: setup.cfg
Comment (sofroniewn) [inline setup.cfg:84]: do we want to make an extra section called "extras" for things that aren't core dependencies, and then include them in "all" too? Maybe that's not needed. Maybe this is also a side note, but should we have a section here called "bundle" too? We could then drop https://github.com/napari/napari/blob/102a7e8f845893c874d2b86f9371d41130100b89/bundle.py#L22-L29 and just use that.

Just thinking about making it really transparent what is being installed where and when. Feel free to say this is something that should be addressed in a followup PR, but just wanted to flag now while on the mind.

File: napari/plugins/_skimage_data.py
Comment (sofroniewn) [reply napari/plugins/_skimage_data.py:42]: makes sense to me. but see some additional comments on details in the .cfg

File: napari/components/viewer_model.py
Comment (sofroniewn) [reply napari/components/viewer_model.py:717]: that's fair. i guess we can punt on figuring out how to expose that in the GUI and just have this like this for now

File: napari/_qt/qt_main_window.py
Comment (sofroniewn) [inline napari/_qt/qt_main_window.py]: How would you feel about stripping the `_` from the names that appear in the dropdown menu, so that a user sees `binary blobs` instead of `binary_blobs`. We might not want to automatically do this for everyone, in which case can you just do it for our builtin skimage plugin. But flagging here in case you did

File: setup.cfg
Comment (sofroniewn) [reply setup.cfg:84]: Do we also want a `>=` version for `scikit-image[data]` too?

**Review by sofroniewn — APPROVED** [review]
Ok, I like the renames, and I'm fine with the kwargs/ adding skimage to napari[all]. I'm approving now, i've left a couple more comments, but all very minor. Thanks for the doc strings too!! We should be able to get this in by tomorrow :-)

File: napari/_qt/qt_main_window.py
Comment (tlambert03) [reply napari/_qt/qt_main_window.py]: I'd probably leave that call to @jni.  It would either mean that we change it in the dict itself (in which case someone would call it programmatically as `viewer.open_sample('scikit-image', 'human mitosis')` but `skimage.data.human_mitosis`) ... or it would mean that we only change it for the menu, in which case they'd have to remember that spaces in the GUI menu sometimes correspond to underscores in the `viewer.open_sample`  call (but possibly not always? depending on the plugin?)

Comment (sofroniewn) [reply napari/_qt/qt_main_window.py]: Hmm i see your points here, but i think most users will only use it in the menu and there having no `_` looks much nicer to me. I'm less worried about the correspondence with `open_sample`, and even less with `skimage.data`. Curious what @jni thinks

Comment (tlambert03) [reply napari/_qt/qt_main_window.py]: yeah, I prefer the space in the menu too.  So, happy to make the change if @jni has no major objections

Comment (goanpeca) [reply napari/_qt/qt_main_window.py]: This is awesome work! I would love to be able to use this, as I am not sure where to look for examples sometimes! 💯 

Now, question. How would this work with localization?

There should be an `ID` `viewer.open_sample('scikit-image', 'human mitosis')` in  this case `human mitosis` but we would also want a human readable display name in the localized language.

If it is the same as the ID then we run into issues with the `viewer.open_sample` since I would not use the localized name for that.

**Comment by sofroniewn** [comment]
I just had one additional thought @tlambert03 could there be away for plugin devs to specify additional metadata about the data beyond the name that we could display in the menubar too. For example in imageJ you see the size of the data, maybe ndim, maybe rgb, maybe shape, maybe layer type? I don't know if it's a good idea or not, but could be something we allow?

Maybe as an optional `name : tuple(func, metadict)`? I dunno, could get complex?

Screen Shot 2021-03-31 at 1 03 54 PM

File: napari/components/viewer_model.py
Comment (goanpeca) [inline napari/components/viewer_model.py]: We need to localize these strings @tlambert03

File: napari/components/viewer_model.py
Comment (goanpeca) [inline napari/components/viewer_model.py]: same here

File: napari/_qt/qt_main_window.py
Comment (sofroniewn) [reply napari/_qt/qt_main_window.py]: Yeah, good point @goanpeca it would be nice to be able to translate these strings too

File: napari/plugins/__init__.py
Comment (goanpeca) [inline napari/plugins/__init__.py:224]: We need to localize these strings @tlambert03

File: napari/plugins/__init__.py
Comment (goanpeca) [inline napari/plugins/__init__.py]: same here

File: napari/components/viewer_model.py
Comment (tlambert03) [reply napari/components/viewer_model.py]: sorry!  it's going to take me a little while to remember that :)  keep at me

File: napari/plugins/_skimage_data.py
Comment (goanpeca) [inline napari/plugins/_skimage_data.py:5]: If we want to have localization of these strings, then we would need to use a mapping.

like https://github.com/napari/napari/blob/master/napari/layers/points/_points_constants.py#L82

File: napari/_qt/qt_main_window.py
Comment (tlambert03) [reply napari/_qt/qt_main_window.py]: > How would this work with localization?

ooof!  😖  very good question!  Since this is plugin-provided, how should we go about thinking about that?

File: napari/plugins/hook_specifications.py
Comment (goanpeca) [inline napari/plugins/hook_specifications.py:56]: Do we need to update some `docs` or that will be done in a separate PR?

**Review by goanpeca — CHANGES_REQUESTED** [review]
Great work @tlambert03 🥳 !

I have some questions regarding localization and cos, otherwise LGTM!

File: napari/plugins/hook_specifications.py
Comment (tlambert03) [reply napari/plugins/hook_specifications.py:56]: This page is what generates the content for https://napari.org/plugins/stable/hook_specifications.html ... so this _is_ the docs... does that cover it?  or did you mean something more narrative?

Comment (goanpeca) [reply napari/plugins/hook_specifications.py:56]: Covers it :)

File: napari/components/viewer_model.py
Comment (goanpeca) [reply napari/components/viewer_model.py]: I have a massive PR with ALL the strings in napari to translate, so soon we will get some CI failures to remind all of us :)

**Comment by neuromusic** [comment]
> a way for plugin devs to specify additional metadata about the data beyond the name

if users need more info than the name, I could imagine a load dialog box rather than a menu entry with thumbnails, metadata, etc

a menu might also be limiting if a user has 10 plugins install each with multiple sample datasets.... could quickly overwhelm the UI

**Comment by neuromusic** [comment]
also just want to add that I think this is super cool and highly valuable for plugin developers and end users

File: napari/_qt/qt_main_window.py
Comment (goanpeca) [reply napari/_qt/qt_main_window.py]: Well plugin devs should handle their own translations, that is they are in charge of doing the whole

`trans._()` on the things they provide. Now I do not know if the hook spec provides a way of having an ID and a "displayName", since this is going to become something we need to provide in several places where we need a unique identifier AND a display version of it 🤔

File: napari/plugins/hook_specifications.py
Comment (goanpeca) [reply napari/plugins/hook_specifications.py:56]: I have no idea how docs work 😆

File: napari/_qt/qt_main_window.py
Comment (tlambert03) [reply napari/_qt/qt_main_window.py]: ok, maybe the spec could be return a dict in addition to a string or a callable?: 

```python
return {
    'sample_id1': 'https://www.img.com/example.png',
    'sample_id2': some_data_function,
    'sample_id3': {
        'display_name': trans._('My Awesome Sample (25K)'),  # would address both localization and display metadata 
        'data': some_data_function,
    }
}
```
?

Comment (sofroniewn) [reply napari/_qt/qt_main_window.py]: that could work!

Comment (goanpeca) [reply napari/_qt/qt_main_window.py]: Works for me :)

File: setup.cfg
Comment (tlambert03) [reply setup.cfg:84]: @jni?

File: napari/components/viewer_model.py
Comment (tlambert03) [reply napari/components/viewer_model.py:717]: actually, a good example of this in the scikit image plugin is:
`viewer.open_sample('scikit-image', 'binary_blobs', length=128, n_dim=3, volume_fraction=0.3)`

Comment (sofroniewn) [reply napari/components/viewer_model.py:717]: yes, that was the one that came to my mind after your comment! I use that one all the time!!

File: setup.cfg
Comment (tlambert03) [reply setup.cfg:84]: made an issue for the bundle extra: https://github.com/napari/napari/issues/2500

File: napari/_qt/qt_main_window.py
Comment (tlambert03) [reply napari/_qt/qt_main_window.py]: updated with localization.  The skimage example shows a bit more detail in the menus now (but not size):

Untitled

File: napari/plugins/__init__.py
Comment (tlambert03) [reply napari/plugins/__init__.py:224]: I'm realizing now that there are a _lot_ of strings in this module and elsewhere that aren't localized (not just this PR).  so, i'm a bit confused as to whether I'm supposed to be doing translation everywhere I see it now?  Or just in the lines added by this PR?

File: napari/plugins/_skimage_data.py
Comment (tlambert03) [reply napari/plugins/_skimage_data.py:5]: have a look at my [updates](https://github.com/napari/napari/pull/2483/files#diff-ab11b76e4d8824d50d5ee4ba9e3dba11ed0a5ea5690cc941ac504640fcafd795R5) and let me if they'll work.

File: napari/plugins/__init__.py
Comment (goanpeca) [reply napari/plugins/__init__.py:224]: I would say just the ones added by this PR. I am working on the rest.

File: napari/components/viewer_model.py
Comment (sofroniewn) [inline napari/components/viewer_model.py]: Should we add `reader_plugin` as a named kwarg here? Could be useful if you know you want to open this sample with this plugin?

File: napari/components/viewer_model.py
Comment (jni) [reply napari/components/viewer_model.py]: > I think I kinda prefer open_sample

Me too.

File: setup.cfg
Comment (jni) [reply setup.cfg:84]: > Do we also want a >= version for scikit-image[data] too?

Huh, I don't know how these things combine... `data` only exists in skimage 0.18 and higher, so does that automatically give us that, or can someone install scikit-image 0.17 and then scikit-image[data] and have scikit-image 0.17 with latest pooch? 🤔

Comment (jni) [reply setup.cfg:84]: Well, that was a mess:

```
 $ pip install scikit-image==0.17.2 scikit-image[data]
Collecting scikit-image==0.17.2
  Using cached scikit_image-0.17.2-cp38-cp38-macosx_10_13_x86_64.whl (12.2 MB)
Collecting scikit-image[data]
  Using cached scikit_image-0.18.1-cp38-cp38-macosx_10_9_x86_64.whl (12.7 MB)
Collecting networkx>=2.0
  Downloading networkx-2.5.1-py3-none-any.whl (1.6 MB)
     |████████████████████████████████| 1.6 MB 5.8 MB/s
Collecting imageio>=2.3.0
  Using cached imageio-2.9.0-py3-none-any.whl (3.3 MB)
Collecting scipy>=1.0.1
  Downloading scipy-1.6.2-cp38-cp38-macosx_10_9_x86_64.whl (30.8 MB)
     |████████████████████████████████| 30.8 MB 11.8 MB/s
Collecting matplotlib!=3.0.0,>=2.0.0
  Downloading matplotlib-3.4.1-cp38-cp38-macosx_10_9_x86_64.whl (7.2 MB)
     |████████████████████████████████| 7.2 MB 9.4 MB/s
Collecting PyWavelets>=1.1.1
  Using cached PyWavelets-1.1.1-cp38-cp38-macosx_10_9_x86_64.whl (4.3 MB)
Collecting numpy>=1.15.1
  Downloading numpy-1.20.2-cp38-cp38-macosx_10_9_x86_64.whl (16.0 MB)
     |████████████████████████████████| 16.0 MB 10.9 MB/s
Collecting pillow!=7.1.0,!=7.1.1,>=4.3.0
  Downloading Pillow-8.2.0-cp38-cp38-macosx_10_10_x86_64.whl (2.8 MB)
     |████████████████████████████████| 2.8 MB 11.0 MB/s
Collecting tifffile>=2019.7.26
  Downloading tifffile-2021.3.31-py3-none-any.whl (165 kB)
     |████████████████████████████████| 165 kB 9.6 MB/s
Collecting cycler>=0.10
  Using cached cycler-0.10.0-py2.py3-none-any.whl (6.5 kB)
Collecting pyparsing>=2.2.1
  Using cached pyparsing-2.4.7-py2.py3-none-any.whl (67 kB)
Collecting python-dateutil>=2.7
  Using cached python_dateutil-2.8.1-py2.py3-none-any.whl (227 kB)
Collecting kiwisolver>=1.0.1
  Using cached kiwisolver-1.3.1-cp38-cp38-macosx_10_9_x86_64.whl (61 kB)
Collecting six
  Using cached six-1.15.0-py2.py3-none-any.whl (10 kB)
Collecting decorator<5,>=4.3
  Using cached decorator-4.4.2-py2.py3-none-any.whl (9.2 kB)
Collecting scikit-image[data]
  Downloading scikit_image-0.18.0-cp38-cp38-macosx_10_9_x86_64.whl (12.7 MB)
     |████████████████████████████████| 12.7 MB 10.3 MB/s
WARNING: scikit-image 0.17.2 does not provide the extra 'data'
Installing collected packages: six, python-dateutil, pyparsing, pillow, numpy, kiwisolver, decorator, cycler, tifffile, scipy, PyWavelets, networkx, matplotlib, imageio, scikit-image
Successfully installed PyWavelets-1.1.1 cycler-0.10.0 decorator-4.4.2 imageio-2.9.0 kiwisolver-1.3.1 matplotlib-3.4.1 networkx-2.5.1 numpy-1.20.2 pillow-8.2.0 pyparsing-2.4.7 python-dateutil-2.8.1 scikit-image-0.17.2 scipy-1.6.2 six-1.15.0 tifffile-2021.3.31

 $ python
Python 3.8.8 (default, Feb 24 2021, 13:46:16)
[Clang 10.0.0 ] :: Anaconda, Inc. on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import skimage
>>> skimage.__version__
'0.17.2'
>>> import pooch
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'pooch'
```

Note the line "WARNING: scikit-image 0.17.2 does not provide the extra 'data'", while at the same time no less than *three* skimage wheels are downloaded for reasons unknown.

tldr, yes, it should be `scikit-image[data]>=0.18`

File: napari/_qt/qt_main_window.py
Comment (jni) [reply napari/_qt/qt_main_window.py]: SO COOL.

happy with the human-readable and localizable names in general. =) I don't love the new nested dictionary but I don't hate it either, and I don't see a better way to do it. So 👍 all around!

**Review by goanpeca — APPROVED** [review]

File: napari/components/viewer_model.py
Comment (tlambert03) [reply napari/components/viewer_model.py]: added this

**Pull request merged**
